File: app/src/main/java/com/example/brl/unitconverter/ConverterFragment.java
package com.example.brl.unitconverter;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;

import static com.example.brl.unitconverter.Constants.MASS_G;
import static com.example.brl.unitconverter.Constants.MASS_OZ;
import static com.example.brl.unitconverter.Constants.TEMP_C;
import static com.example.brl.unitconverter.Constants.TEMP_F;
import static com.example.brl.unitconverter.Constants.TYPE_CURRENCY;
import static com.example.brl.unitconverter.Constants.TYPE_DISTANCE;
import static com.example.brl.unitconverter.Constants.TYPE_MASS;
import static com.example.brl.unitconverter.Constants.TYPE_TEMPERATURE;

public class ConverterFragment extends Fragment implements AdapterView.OnItemSelectedListener {

    Spinner typeSpinner;
    Spinner inputSpinner;
    Spinner outputSpinner;
    ArrayAdapter<CharSequence> typeAdapter;
    ArrayAdapter<CharSequence> massAdapter;
    ArrayAdapter<CharSequence> temperatureAdapter;
    ArrayAdapter<CharSequence> distanceAdapter;
    ArrayAdapter<CharSequence> currencyAdapter;

    EditText inputEditText;
    EditText outputEditText;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_converter, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        typeSpinner = (Spinner)view.findViewById(R.id.fragment_converter_type_spinner);
        inputSpinner = (Spinner)view.findViewById(R.id.fragment_converter_input_spinner);
        outputSpinner = (Spinner)view.findViewById(R.id.fragment_converter_output_spinner);
        inputEditText = (EditText)view.findViewById(R.id.fragment_converter_input_editText);
        outputEditText = (EditText)view.findViewById(R.id.fragment_converter_output_editText);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        setupAdapters();

        setupSpinner(typeSpinner, typeAdapter);
        setupSpinner(inputSpinner, massAdapter);
        setupSpinner(outputSpinner, massAdapter);
        setupTextWatchers();
    }


    private void setupTextWatchers() {
        inputEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                if (inputEditText.hasFocus()) {
                    parametersChanged(true);
                }
            }
        });
        outputEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                if (outputEditText.hasFocus()) {
                    parametersChanged(false);
                }
            }
        });
    }

    private void setupAdapters() {
        typeAdapter = ArrayAdapter.createFromResource(getContext(), R.array.type_array, android.R.layout.simple_spinner_item);
        typeAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        massAdapter = ArrayAdapter.createFromResource(getContext(), R.array.mass_array, android.R.layout.simple_spinner_item);
        massAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        temperatureAdapter = ArrayAdapter.createFromResource(getContext(), R.array.temperature_array, android.R.layout.simple_spinner_item);
        temperatureAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        distanceAdapter = ArrayAdapter.createFromResource(getContext(), R.array.distance_array, android.R.layout.simple_spinner_item);
        distanceAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        currencyAdapter = ArrayAdapter.createFromResource(getContext(), R.array.currency_array, android.R.layout.simple_spinner_item);
        currencyAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        switch (parent.getId()) {
            case R.id.fragment_converter_type_spinner:
                handleTypeSelected(position);
                break;
            case R.id.fragment_converter_input_spinner:
                parametersChanged(true);
                break;
            case R.id.fragment_converter_output_spinner:
                parametersChanged(false);
                break;
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {

    }

    private void setupSpinner(Spinner spinner, ArrayAdapter adapter) {
        spinner.setAdapter(adapter);
        spinner.setOnItemSelectedListener(this);
    }

    private void handleTypeSelected(int position) {
        switch (position) {
            case TYPE_MASS:
                inputSpinner.setAdapter(massAdapter);
                inputSpinner.setSelection(MASS_G);
                outputSpinner.setAdapter(massAdapter);
                outputSpinner.setSelection(MASS_OZ);
                break;
            case TYPE_TEMPERATURE:
                inputSpinner.setAdapter(temperatureAdapter);
                inputSpinner.setSelection(TEMP_C);
                outputSpinner.setAdapter(temperatureAdapter);
                outputSpinner.setSelection(TEMP_F);
                break;
            case TYPE_DISTANCE:
                break;
            case TYPE_CURRENCY:
                break;
        }
    }

    private void parametersChanged(boolean inputChanged) {
        switch (typeSpinner.getSelectedItemPosition()) {
            case TYPE_MASS:
                if (inputChanged) {
                    float input = Utils.toGrams(inputSpinner.getSelectedItemPosition(), Utils.getFloatFromString(inputEditText.getText().toString()));
                    float output = Utils.fromGrams(outputSpinner.getSelectedItemPosition(), input);
                    outputEditText.setText(String.valueOf(output));
                } else {
                    float input = Utils.toGrams(outputSpinner.getSelectedItemPosition(), Utils.getFloatFromString(outputEditText.getText().toString()));
                    float output = Utils.fromGrams(inputSpinner.getSelectedItemPosition(), input);
                    inputEditText.setText(String.valueOf(output));
                }
                break;
            case TYPE_TEMPERATURE:
                break;
            case TYPE_DISTANCE:
                break;
            case TYPE_CURRENCY:
                break;
        }
    }
}

File: app/src/main/java/com/example/brl/unitconverter/ConverterFragment.kt
package com.example.brl.unitconverter

import android.os.Bundle
import android.support.v4.app.Fragment
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.Spinner

import kotlinx.android.synthetic.main.fragment_converter.*

import com.example.brl.unitconverter.Constants.MASS_G
import com.example.brl.unitconverter.Constants.MASS_OZ
import com.example.brl.unitconverter.Constants.TEMP_C
import com.example.brl.unitconverter.Constants.TEMP_F
import com.example.brl.unitconverter.Constants.TYPE_CURRENCY
import com.example.brl.unitconverter.Constants.TYPE_DISTANCE
import com.example.brl.unitconverter.Constants.TYPE_MASS
import com.example.brl.unitconverter.Constants.TYPE_TEMPERATURE

class ConverterFragment : Fragment(), AdapterView.OnItemSelectedListener {

//    internal var typeSpinner: Spinner? = null
//    internal var inputSpinner: Spinner? = null
//    internal var outputSpinner: Spinner? = null
    internal var typeAdapter: ArrayAdapter<CharSequence>? = null
    internal var massAdapter: ArrayAdapter<CharSequence>? = null
    internal var temperatureAdapter: ArrayAdapter<CharSequence>? = null
    internal var distanceAdapter: ArrayAdapter<CharSequence>? = null
    internal var currencyAdapter: ArrayAdapter<CharSequence>? = null

//    internal var inputEditText: EditText? = null
//    internal var outputEditText: EditText? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_converter, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        setupAdapters()

        setupSpinner(fragment_converter_type_spinner, typeAdapter)
        setupSpinner(fragment_converter_input_spinner, massAdapter)
        setupSpinner(fragment_converter_output_spinner, massAdapter)
        setupTextWatchers()
    }


    private fun setupTextWatchers() {
        fragment_converter_input_editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

            }

            override fun afterTextChanged(s: Editable) {
                if (fragment_converter_input_editText.hasFocus()) {
                    parametersChanged(true)
                }
            }
        })
        fragment_converter_output_editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

            }

            override fun afterTextChanged(s: Editable) {
                if (fragment_converter_output_editText.hasFocus()) {
                    parametersChanged(false)
                }
            }
        })
    }

    private fun setupAdapters() {
        typeAdapter = ArrayAdapter.createFromResource(context, R.array.type_array, android.R.layout.simple_spinner_item)
        typeAdapter?.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        massAdapter = ArrayAdapter.createFromResource(context, R.array.mass_array, android.R.layout.simple_spinner_item)
        massAdapter?.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        temperatureAdapter = ArrayAdapter.createFromResource(context, R.array.temperature_array, android.R.layout.simple_spinner_item)
        temperatureAdapter?.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        distanceAdapter = ArrayAdapter.createFromResource(context, R.array.distance_array, android.R.layout.simple_spinner_item)
        distanceAdapter?.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        currencyAdapter = ArrayAdapter.createFromResource(context, R.array.currency_array, android.R.layout.simple_spinner_item)
        currencyAdapter?.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
    }

    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        when (parent?.id) {
            R.id.fragment_converter_type_spinner -> handleTypeSelected(position)
            R.id.fragment_converter_input_spinner -> parametersChanged(true)
            R.id.fragment_converter_output_spinner -> parametersChanged(false)
        }
    }

    override fun onNothingSelected(parent: AdapterView<*>) {

    }

    private fun setupSpinner(spinner: Spinner, adapter: ArrayAdapter<CharSequence>?) {
        spinner.adapter = adapter
        spinner.onItemSelectedListener = this
    }

    private fun handleTypeSelected(position: Int) {
        when (position) {
            TYPE_MASS -> {
                fragment_converter_input_spinner.adapter = massAdapter
                fragment_converter_input_spinner.setSelection(MASS_G)
                fragment_converter_output_spinner.adapter = massAdapter
                fragment_converter_output_spinner.setSelection(MASS_OZ)
            }
            TYPE_TEMPERATURE -> {
                fragment_converter_input_spinner.adapter = temperatureAdapter
                fragment_converter_input_spinner.setSelection(TEMP_C)
                fragment_converter_output_spinner.adapter = temperatureAdapter
                fragment_converter_output_spinner.setSelection(TEMP_F)
            }
            TYPE_DISTANCE -> {
            }
            TYPE_CURRENCY -> {
            }
        }
    }

    private fun parametersChanged(inputChanged: Boolean) {
        when (fragment_converter_type_spinner.selectedItemPosition) {
            TYPE_MASS -> if (inputChanged) {
                val input = Utils.toGrams(fragment_converter_input_spinner.selectedItemPosition, Utils.getFloatFromString(fragment_converter_input_editText.text.toString()))
                val output = Utils.fromGrams(fragment_converter_output_spinner.selectedItemPosition, input)
                fragment_converter_output_editText.setText(output.toString())
            } else {
                val input = Utils.toGrams(fragment_converter_output_spinner.selectedItemPosition, Utils.getFloatFromString(fragment_converter_output_editText.text.toString()))
                val output = Utils.fromGrams(fragment_converter_input_spinner.selectedItemPosition, input)
                fragment_converter_input_editText.setText(output.toString())
            }
            TYPE_TEMPERATURE -> {
            }
            TYPE_DISTANCE -> {
            }
            TYPE_CURRENCY -> {
            }
        }
    }
}
